homeassistant/components/esphome/config_flow.py

```diff
@@ -288,14 +288,56 @@ async def async_step_mqtt(
 
         return await self.async_step_discovery_confirm()
 
+    async def try_connect(self, entry: ConfigEntry) -> str | None:
+        """Try connecting to a device and return any errors."""
+        host = entry.data[CONF_HOST]
+        port = entry.data[CONF_PORT]
+        password = entry.data[CONF_PASSWORD]
+        psk = entry.data.get(CONF_NOISE_PSK, "")
+
+        zeroconf_instance = await zeroconf.async_get_instance(self.hass)
+
+        cli = APIClient(
+            host,
+            port,
+            password,
+            zeroconf_instance=zeroconf_instance,
+            noise_psk=psk,
+        )
+
+        try:
+            await cli.connect()
+        except APIConnectionError:
+            return "connection_error"
+        finally:
+            await cli.disconnect(force=True)
+
```

Comment:
This code is substantially similar to `fetch_device_info`. It would be good to split `fetch_device_info` into two functions so you can reuse part of it here.

Reply:
Yeah, I thought about it. But the similarity is really only in three minor aspects:

1) acquiring zeroconf instance
2) creating APIClient
3) calling connect on APIClient instance

`fetch_device_info` (as well as `try_login`) differ from `try_connect` in where they obtain their data from as well as how they respond to various exceptions and failures. I can extract the three above mentioned parts into their own method, but it's not going to be as significant as one might expect at first.